vertx-web-client/src/main/java/io/vertx/ext/web/client/impl/cache/CacheKey.java

```diff
@@ -0,0 +1,85 @@
+/*
+ * Copyright 2021 Red Hat, Inc.
+ *
+ *  All rights reserved. This program and the accompanying materials
+ *  are made available under the terms of the Eclipse Public License v1.0
+ *  and Apache License v2.0 which accompanies this distribution.
+ *
+ *  The Eclipse Public License is available at
+ *  http://www.eclipse.org/legal/epl-v10.html
+ *
+ *  The Apache License v2.0 is available at
+ *  http://www.opensource.org/licenses/apache2.0.php
+ *
+ *  You may elect to redistribute this code under either of these licenses.
+ */
+package io.vertx.ext.web.client.impl.cache;
+
+import io.vertx.ext.web.client.HttpRequest;
+import io.vertx.ext.web.client.spi.CacheStore;
+import java.nio.charset.StandardCharsets;
+import java.security.MessageDigest;
+import java.util.Objects;
+
+/**
+ * A key for a {@link CacheStore} based on a {@link HttpRequest}.
+ *
+ * @author <a href="mailto:craigday3@gmail.com">Craig Day</a>
+ */
+public class CacheKey extends CacheVariationsKey {
+
+  private final String variations;
+
+  public CacheKey(HttpRequest<?> request, Vary vary) {
+    super(request);
+    this.variations = vary.toString();
+  }
+
+  @Override
+  public String toString() {
+    String rawKey = super.toString() + "|" + variations;
```

Comment:
we should avoid concatenating then computing the digest and instead compute the digest of the three strings:

```java
digest.update(super.toString().getBytes(StandardCharsets.UTF_8));
digest.update("|".getBytes(StandardCharsets.UTF_8));
digest.update(variations.getBytes(StandardCharsets.UTF_8));
byte[] hashed = digest.digest();
```

I think we should even try to do something more optimized here with a faster algorithm here and avoid using a toString()

Reply:
is that actually called often or use for comparison ?

Reply:
i.e do we care about performance here ?

Reply:
That's partly an implementation detail of the CacheStore. If you're using a remote store like redis or memcached, then you'll want a nice concise string to use as the key. In that case, this will get called for each get/set. I generally prefer opaque keys for cases like this. I could probably use `Objects.hash` instead of a digest library if we're concerned about performance here.